src/families/cosmos/react.js

```diff
@@ -14,3 +24,143 @@ export function useCosmosPreloadData() {
 }
+
+export function useCosmosFormattedDelegations(
+  account: Account,
+  option?: FormatOption
+): CosmosFormattedDelegation[] {
+  const { validators } = useCosmosPreloadData();
+  const delegations = account.cosmosResources?.delegations;
+  invariant(delegations, "cosmos: delegations is required");
+
+  const formattedDelegations = useMemo(
+    () => formatDelegations(delegations, validators),
+    [delegations, validators]
+  );
+  const unit = useMemo(() => getAccountUnit(account), [account]);
+
+  switch (option) {
+    case "claimReward":
+      return formattedDelegations
+        .filter(({ pendingRewards }) => pendingRewards.gt(0))
+        .map(({ pendingRewards, ...rest }) => ({
+          ...rest,
+          pendingRewards,
+          reward: formatCurrencyUnit(unit, pendingRewards, {
+            disableRounding: true,
+            alwaysShowSign: false,
+            showCode: true,
+          }),
+        }));
+    case "redelegate":
+    case "undelegate":
+      return formattedDelegations.map((d) => ({
+        ...d,
+        formattedAmount: formatCurrencyUnit(unit, d.amount, {
+          disableRounding: true,
+          alwaysShowSign: false,
+          showCode: true,
+        }),
+      }));
+    default:
+      return formattedDelegations;
+  }
+}
+
+export function useCosmosDelegationsQuerySelector(
+  account: Account,
+  transaction: Transaction,
+  option?: FormatOption
+) {
+  const [query, setQuery] = useState<string>("");
+  const delegations = useCosmosFormattedDelegations(account, option);
+
+  const options = useMemo<CosmosFormattedDelegation[]>(
+    () =>
+      delegations.filter(
+        // [TODO] better query test
+        ({ validator }) =>
+          !query || !validator || new RegExp(query, "gi").test(validator.name)
+      ),
+    [query, delegations]
+  );
+
+  const selectedValidator = useMemo(
+    () => transaction.validators && transaction.validators[0],
+    [transaction]
+  );
+
+  const value = useMemo(() => {
+    switch (option) {
+      case "redelegate":
+        invariant(
+          transaction.cosmosSourceValidator,
+          "cosmos: cosmosSourceValidator is required"
+        );
+        return options.find(
+          ({ address }) => address === transaction.cosmosSourceValidator
+        );
+      default:
+        return (
+          selectedValidator &&
+          delegations.find(
+            ({ address }) => address === selectedValidator.address
+          )
+        );
+    }
+  }, [delegations, selectedValidator, transaction, options, option]);
+
+  return {
+    query,
+    setQuery,
+    options,
+    value,
+  };
+}
+
+/** Hook to search and sort SR list according to initial votes and query */
+export function useSortedValidators(
+  search: string,
+  validators: CosmosValidatorItem[],
+  delegations: CosmosDelegationInfo[]
+): {
+  validator: CosmosValidatorItem,
+  name: ?string,
+  address: string,
+  rank: number,
+}[] {
+  const { current: initialVotes } = useRef(
```

Comment:
the use of `useRef` here is suboptimal, you are always calculating `delegations.map(({ address }) => address)` here and pass it to useRef. personally i would just have use `const [initialVotes] = useState(() => delegations.map(({ address }) => address))` that (even if it's a read only state) is clearer to read and also it benefit from fact the init-er is called once.

Reply:
@LFBarreto

Reply:
i'm not sure the useRef recalculates but sure we can do this with useState instead might be clearer to read :+1:

Reply:
@gre In this case, useMemo with empty dependencies is more appropriate imo.
@LFBarreto I checked whether useRef recalculates and yes, it does.

Reply:
since useMemo doesn't guarantee to prevent re-computation, I would suggest to create useMemoOnce.

```
function useMemoOnce<T>(callback: () => T): T {
  const ref = useRef()
 
  if (!ref.current) {
    ref.current = callback()
  }

  return ref.current
}
```